### PR TITLE
Trap panic in GetIndexEntry and log an ERROR

### LINKED ISSUE #2777: Prevent panic in rollbackEntries and log debug info
This ticket has been split out from #2605 to prevent the panic and log debug information.

### PR DESCRIPTION
**Pull request by ajres**
fixes #2777

### CODE REVIEW

File: db/kv_dense_block.go [hunk 1]
```diff
@@ -594,7 +595,10 @@ func (d *DenseBlock) MakeLogEntry(indexEntry DenseBlockIndexEntry, entry DenseBl
 }
 
 func (d *DenseBlock) GetIndexEntry(position int64) (indexEntry DenseBlockIndexEntry) {
-	indexEntry = d.value[position : position+INDEX_ENTRY_LEN]
+	indexEntry, err := base.SafeSlice(d.value, int(position), int(position+INDEX_ENTRY_LEN))
+	if err != nil {
+		base.LogError(errors.New(fmt.Sprintf("Unable to GetIndexEntry from DensBlockIndexEntry, key: %v, cas: %v, startClock: %v, clock: %v",d.Key, d.cas, d.startClock, d.clock)))
```
Comment (adamcfraser) [inline db/kv_dense_block.go]: I think it's useful to include the Key here, and possibly the cas.  The clocks aren't going to be logged as anything useful, I don't think (just pointers), and I think it's critical to log the error text that's being returned from SafeSlice.

**Review by adamcfraser — CHANGES_REQUESTED** [review]

File: db/kv_dense_block.go [hunk 2]
```diff
@@ -594,7 +595,10 @@ func (d *DenseBlock) MakeLogEntry(indexEntry DenseBlockIndexEntry, entry DenseBl
 }
 
 func (d *DenseBlock) GetIndexEntry(position int64) (indexEntry DenseBlockIndexEntry) {
-	indexEntry = d.value[position : position+INDEX_ENTRY_LEN]
+	indexEntry, err := base.SafeSlice(d.value, int(position), int(position+INDEX_ENTRY_LEN))
```
Comment (adamcfraser) [inline db/kv_dense_block.go:602]: I think there are a couple more places where we do slice targeting in dense_block - it feels like we should at least switch over GetEntry at the same time.

Comment (ajres) [reply db/kv_dense_block.go:602]: Added SafeSlice to GetEntry

Comment (ajres) [reply db/kv_dense_block.go:602]: Updated log message format to remove clocks and add SafeSlice error message

**Comment by ajres** [comment]
Commented out the block in rollbackEntries that checks for a nil indexEntry and re-ran the unit tests. The new empty block rollback test failed:

```
2017-08-10T20:35:25.811+01:00 ERROR: Unable to GetIndexEntry from DensBlock, key: block2, cas: 0, error: Invalid slice [786422:786434] of []byte with len 2 -- db.(*DenseBlock).GetIndexEntry() at kv_dense_block.go:604
--- FAIL: TestDenseBlockRollbackTo (0.00s)
panic: runtime error: slice bounds out of range [recovered]
	panic: runtime error: slice bounds out of range

goroutine 2029 [running]:
panic(0x56f3a0, 0xc420012100)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
testing.tRunner.func1(0xc420da4240)
	/usr/local/go/src/testing/testing.go:579 +0x25d
panic(0x56f3a0, 0xc420012100)
	/usr/local/go/src/runtime/panic.go:458 +0x243
github.com/couchbase/sync_gateway/db.(*DenseBlock).rollbackEntries(0xc421f37df8, 0xc420da0001, 0x1, 0x0, 0x2b)
	/Users/andy/Projects/sync_gateway-2777/godeps/src/github.com/couchbase/sync_gateway/db/kv_dense_block.go:376 +0x1c5
github.com/couchbase/sync_gateway/db.(*DenseBlock).RollbackTo(0xc421f37df8, 0xc420da0001, 0x1, 0x8dc880, 0xc42093a880, 0xc4213bd200, 0x0, 0x0)
	/Users/andy/Projects/sync_gateway-2777/godeps/src/github.com/couchbase/sync_gateway/db/kv_dense_block.go:303 +0x6e
github.com/couchbase/sync_gateway/db.TestDenseBlockRollbackTo(0xc420da4240)
	/Users/andy/Projects/sync_gateway-2777/godeps/src/github.com/couchbase/sync_gateway/db/kv_dense_channel_storage_test.go:387 +0x1f87
testing.tRunner(0xc420da4240, 0x6b65b0)
	/usr/local/go/src/testing/testing.go:610 +0x81
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:646 +0x2ec
FAIL	github.com/couchbase/sync_gateway/db	17.303s
```

**Review by adamcfraser — APPROVED** [review]

**Comment by adamcfraser** [comment]
Looks good.  @ajres Can you rebase to address the conflicts, and then we can merge?

**Pull request merged**